### PR TITLE
Configure multiple notification channels

### PR DESCRIPTION
**Pull request by DavidKang**
Nowadays, we have three kinds of channels (rss, email and web), but they aren't configurable.
With this PR we want to provide to the user the possibility to manage their notification channels.

### Large
Screenshot_2020-04-16 Open Build Service(1)

### Medium
Screenshot_2020-04-17 Open Build Service

### Small
Screenshot_2020-04-17 Open Build Service(1)

### CODE REVIEW

**Comment by hennevogel** [comment]
If you come to this page a month after you configured it the first time. 

Screenshot from 2020-04-16 19-04-51

What is on and what is off? Also the space on the right of the buttons looks weird. In other places we use the switch, see https://obs-patterns.netlify.app/molecules/forms.html

**Comment by DavidKang** [comment]
> What is on and what is off? 

IMO, is clear, but maybe because I already know it. A full colored button is on/enable.

> Also the space on the right of the buttons looks weird.

The space on the right is there because we have a spinner to tell to the user that something is happening. I can remove this space, but then when the spinner turns visible the box will grow a little bit and move the content. That IMO, is worse.

 > In other places we use the switch, see https://obs-patterns.netlify.app/molecules/forms.html

I don't mind to use the switch, but then I will have to re-think again how to distribute the content.
I will try it with the "switch".

**Comment by hennevogel** [comment]
> I will try it with the "switch".

Read up on when to use what first please

**Comment by DavidKang** [comment]
> Read up on when to use what first please

Then I will discard the `switch`. It doesn't apply in this case.

**Comment by DavidKang** [comment]
@hennevogel

Screenshot_2020-04-16 Open Build Service(1)

**Comment by hennevogel** [comment]
Yeah I guess this is the way :smile:

**Comment by DavidKang** [comment]
I splitted this PR in two, to make it easier to review. This part is only about, the notification configuration. I will created a following PR were we adapt the job and the user notifications controller.

File: src/api/app/views/webui/subscriptions/_subscriptions_form.html.haml [hunk 1]
```diff
@@ -1,17 +1,21 @@
 - i = 0
-- subscriptions_form.subscriptions_by_event.each do |event_class, subscriptions|
+- subscriptions_form.subscriptions_by_event.each do |subscription_by_event|
```
Comment (hennevogel) [inline src/api/app/views/webui/subscriptions/_subscriptions_form.html.haml:2]:
```suggestion
- subscriptions_form.subscriptions_by_event.each_with_index do |subscription|
```

Comment (hennevogel) [reply src/api/app/views/webui/subscriptions/_subscriptions_form.html.haml:2]: Then you can also skip the `i`...

File: src/api/app/models/event_subscription/for_role_form.rb [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+    def find_subscription_for_event_class_and_role(event_class, role, channel)
+      subscriber_subscription =
+        find_subscription_for_subscriber(event_class, role, channel) || find_subscription_for_subscriber(event_class, :all, channel)
+      default_subscription = find_default_subscription(event_class, role, channel)
+
+      # 1. Pick the subscriber's subscription if it exists
+      subscription = if subscriber.present? && subscriber_subscription.present?
+                       subscriber_subscription
+
+                     # 2. Pick the subscriber's subscription if it exists
```
Comment (hennevogel) [inline src/api/app/models/event_subscription/for_role_form.rb]:
```suggestion
                     # 2. Pick the default subscription if it exists
```

File: src/api/app/models/event_subscription/for_role_form.rb [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+      default_subscriptions.find { |s| s.event_class == event_class && s.receiver_role == role && s.channel == channel }
+    end
+
+    def subscriber_subscriptions
+      @subscriber_subscriptions ||= EventSubscription.for_subscriber(subscriber)
+    end
+
+    def default_subscriptions
+      @default_subscriptions ||= EventSubscription.defaults
+    end
```
Comment (hennevogel) [inline src/api/app/models/event_subscription/for_role_form.rb:63]: Hm what do we gain by caching `subscriber_subscriptions` and `default_subscriptions` in instance variables? Does this have any measurable impact?

File: src/api/app/models/event_subscription/for_role_form.rb [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+
+                     # 3. Otherwise instantiate a new subscription
+                     else
+                       EventSubscription.new(
+                         subscriber: subscriber,
+                         eventtype: event_class.to_s,
+                         receiver_role: role,
+                         channel: 'disabled'
+                       )
+                     end
```
Comment (hennevogel) [inline src/api/app/models/event_subscription/for_role_form.rb:44]: I don't think those comments add any additional information while making the formatting hard to parse. Do we really need them?

File: src/api/app/controllers/webui/users/subscriptions_controller.rb [hunk 5]
```diff
@@ -20,9 +20,9 @@ def update
     end
 
     subscriptions_form.update!(params[:subscriptions]) if params[:subscriptions]
-    flash[:success] = 'Notifications settings updated'
+    flash.now[:success] = 'Notifications settings updated'
   rescue ActiveRecord::RecordInvalid
-    flash[:error] = 'Notifications settings could not be updated due to an error'
+    flash.now[:error] = 'Notifications settings could not be updated due to an error'
```
Comment (hennevogel) [inline src/api/app/controllers/webui/users/subscriptions_controller.rb:25]: What error?

File: src/api/app/models/event_subscription/for_role_form.rb [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+    end
+
+    private
+
+    def find_subscription_for_event_class_and_role(event_class, role, channel)
+      subscriber_subscription =
+        find_subscription_for_subscriber(event_class, role, channel) || find_subscription_for_subscriber(event_class, role, 'disabled')
+      default_subscription = find_default_subscription(event_class, role, channel)
+
+      subscription = if subscriber.present? && subscriber_subscription.present?
```
Comment (danidoni) [inline src/api/app/models/event_subscription/for_role_form.rb:27]: I think you can write something like this
```
subscription = subscriber_subscription if subscriber.present?
subscription ||= default_subscription
subscription ||= EventSubscription.new(subscriber: subscriber,
                                       eventtype: event_class.to_s,
                                       receiver_role: role,
                                       channel: 'disabled')
```
```

File: src/api/app/models/event_subscription/for_role_form.rb (hunk 3 above)
Comment (danidoni) [reply src/api/app/models/event_subscription/for_role_form.rb:57]: The form is called only once so I guess not...

File: src/api/app/controllers/webui/users/subscriptions_controller.rb [hunk 7]
```diff
@@ -20,9 +20,9 @@ def update
     end
 
     subscriptions_form.update!(params[:subscriptions]) if params[:subscriptions]
-    flash[:success] = 'Notifications settings updated'
-  rescue ActiveRecord::RecordInvalid
-    flash[:error] = 'Notifications settings could not be updated due to an error'
+    flash.now[:success] = 'Notifications settings updated'
+  rescue ActiveRecord::RecordInvalid => e
```
Comment (dmarcoux) [inline src/api/app/controllers/webui/users/subscriptions_controller.rb:24]: While we're at it, let's use `error` instead of `e`.

File: src/api/app/models/event_subscription/for_role_form.rb (hunk 4 above)
Comment (DavidKang) [reply src/api/app/models/event_subscription/for_role_form.rb:38]: I removed the comments

File: src/api/app/models/event_subscription/for_role_form.rb (hunk 6 above)
Comment (DavidKang) [reply src/api/app/models/event_subscription/for_role_form.rb:27]: I will have a look on it

File: src/api/app/models/event_subscription/for_event_form.rb [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+    attr_reader :event_class, :subscriber, :roles
+
+    def initialize(event, subscriber)
+      @subscriber = subscriber
+      @event_class = event
+      @roles = []
+    end
+
+    def call
+      @roles = @event_class.receiver_roles.map { |role| EventSubscription::ForRoleForm.new(role, event_class, subscriber).call }
```
Comment (dmarcoux) [inline src/api/app/models/event_subscription/for_event_form.rb:12]: Since there's `attr_reader :event_class`, we should use `event_class`.

Comment (DavidKang) [reply src/api/app/models/event_subscription/for_event_form.rb]: Done!

File: src/api/app/controllers/webui/users/subscriptions_controller.rb (hunk 7 above)
Comment (DavidKang) [reply src/api/app/controllers/webui/users/subscriptions_controller.rb:24]: Rubocop prefers `e` instead of `error`

File: src/api/app/models/event_subscription/for_role_form.rb (hunk 6 above)
Comment (DavidKang) [reply src/api/app/models/event_subscription/for_role_form.rb]: I did it in a different way.

File: src/api/app/controllers/webui/users/subscriptions_controller.rb (hunk 5 above)
Comment (DavidKang) [reply src/api/app/controllers/webui/users/subscriptions_controller.rb]: Done

File: src/api/app/controllers/webui/users/subscriptions_controller.rb (hunk 7 above)
Comment (dmarcoux) [reply src/api/app/controllers/webui/users/subscriptions_controller.rb:24]: Oh, because of [Naming/RescuedExceptionsVariableName](https://rubocop.readthedocs.io/en/latest/cops_naming/#namingrescuedexceptionsvariablename). I'm not a fan of this cop, but whatever, I don't care enough to change this setting from the default value.

File: src/api/app/models/event_subscription/for_role_form.rb (hunk 3 above)
Comment (DavidKang) [reply src/api/app/models/event_subscription/for_role_form.rb:52]: Even if the impact is not so big, I would prefer to keep the memoize of both method. We have a performance of 50% aprox.


|  | user | system | total | real |
| --- | --- | --- | --- | --- |
| memoized | 0.116647  | 0.000000 | 0.116647 | (  0.123458) |
| no memoized | 0.213298 | 0.010817 | 0.224115 | (  0.252572)|

File: src/api/app/views/webui/subscriptions/_subscriptions_form.html.haml (hunk 1 above)
Comment (DavidKang) [reply src/api/app/views/webui/subscriptions/_subscriptions_form.html.haml:2]: In this case, we can't use `each_with_index` because the index will be reset in each event, and we don't want that.

File: src/api/app/models/event_subscription/for_role_form.rb (hunk 6 above)
Comment (danidoni) [reply src/api/app/models/event_subscription/for_role_form.rb]: Okay

File: src/api/app/views/webui/subscriptions/_subscriptions_form.html.haml (hunk 1 above)
Comment (hennevogel) [reply src/api/app/views/webui/subscriptions/_subscriptions_form.html.haml:2]: nitpick: It would still help readability if the variable would be called `subscription`

File: src/api/app/models/event_subscription/for_role_form.rb [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+
+    def initialize(role_name, event, subscriber)
+      @subscriber = subscriber
+      @name = role_name
+      @event = event
+      @channels = []
+    end
+
+    def call
+      @channels = EventSubscription.channels.keys[1..-1].map do |channel|
```
[the remark this answers is not in the document]
Comment (danidoni) [reply src/api/app/models/event_subscription/for_role_form.rb:13]: What about this `EventSubscription.channels.keys.drop(1).reject {|k,v| k == "disabled"}.map do |channel|`.

[the remark this answers is not in the document]
Comment (DavidKang) [reply src/api/app/models/event_subscription/for_role_form.rb]: changed

File: src/api/app/views/webui/subscriptions/_subscriptions_form.html.haml (hunk 1 above)
Comment (DavidKang) [reply src/api/app/views/webui/subscriptions/_subscriptions_form.html.haml]: Changed

**Comment by DavidKang** [comment]
DO NOT MERGER, because I found a bug.

File: src/api/app/models/event_subscription/for_role_form.rb [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+
+    def initialize(role_name, event, subscriber)
+      @subscriber = subscriber
+      @name = role_name
+      @event = event
+      @channels = []
+    end
+
+    def call
+      @channels = EventSubscription.channels.keys.reject { |channel| channel == 'disabled' }.map do |channel|
```
Comment (danidoni) [inline src/api/app/models/event_subscription/for_role_form.rb:13]: :thought_balloon: To** maket it shorter, you can extract this into a `EventSubscription#channels_without_disabled` :stuck_out_tongue:

Comment (DavidKang) [reply src/api/app/models/event_subscription/for_role_form.rb:13]: right!, thanks :smile_cat:
